protobuf/src/main/scala/magnolify/protobuf/ProtobufType.scala

```diff
@@ -89,16 +91,29 @@ object ProtobufField {
   type Typeclass[T] = ProtobufField[T]
 
   def combine[T](caseClass: CaseClass[Typeclass, T]): Record[T] = new Record[T] {
+    // One Record[T] instance may be used for multiple Message types
+    @transient private val fieldsCache: concurrent.Map[String, Array[FieldDescriptor]] =
+      concurrent.TrieMap.empty
+    private def getFields(descriptor: Descriptor): Array[FieldDescriptor] =
+      fieldsCache.getOrElseUpdate(
+        descriptor.getFullName, {
+          val fields = new Array[FieldDescriptor](caseClass.parameters.size)
+          caseClass.parameters.foreach(p => fields(p.index) = descriptor.findFieldByName(p.label))
+          fields
+        }
+      )
+
     override val hasOptional: Boolean = caseClass.parameters.exists(_.typeclass.hasOptional)
 
     override def from(v: Message): T = {
       val descriptor = v.getDescriptorForType
       val syntax = descriptor.getFile.getSyntax
+      val fields = getFields(descriptor)
 
       caseClass.construct { p =>
-        val field = descriptor.findFieldByName(p.label)
```

Comment:
Looks like `findFieldByName` is O(n) internally so this would make a bigger difference for fat objects.